### ISSUE TITLE
argdist error messages and includes

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bcc#374]
```
# ./argdist.py -C 'p:md_end_flush(struct bio *bio):int:bio->bi_private'
Traceback (most recent call last):
  File "./argdist.py", line 255, in <module>
    specifiers.append(Specifier("freq", specifier, args.pid))
  File "./argdist.py", line 58, in __init__
    raise ValueError("invalid specifier format")
ValueError: invalid specifier format
```

That's a typo, should be "p::", but the error message could be improved.

Then:

```
# ./argdist.py -C 'p::md_end_flush(struct bio *bio):int:bio->bi_private'
<bcc-memory-buffer>:5:53: warning: declaration of 'struct bio' will not be visible outside of this function [-Wvisibility]
int md_end_flush_probe0(struct pt_regs *ctx ,struct bio *bio)
                                                    ^
<bcc-memory-buffer>:8:24: error: incomplete definition of type 'struct bio'
        int __key = bio->bi_private;
                    ~~~^
<bcc-memory-buffer>:5:53: note: forward declaration of 'struct bio'
int md_end_flush_probe0(struct pt_regs *ctx ,struct bio *bio)
                                                    ^
1 warning and 1 error generated.
Traceback (most recent call last):
  File "./argdist.py", line 269, in <module>
    bpf = BPF(text=bpf_source)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 483, in __init__
    raise Exception("Failed to compile BPF module %s" % src_file)
Exception: Failed to compile BPF module 
```

I can fix this by editing argdist.py and adding a "#include \<linux/blkdev.h\>". Perhaps it needs a way to specify this on the command line, or, perhaps this is getting out of the scope for the tool.

**Comment by goldshtn** [comment]
I will add better error reporting for invalid probe syntax. Regarding the structure definitions, one of my examples shows that you can declare the struct inline as part of the signature:

```
# argdist.py -C 'p:c:nanosleep(struct timespec { time_t tv_sec; long tv_nsec; } *req):long:req->tv_nsec'
```

It doesn't actually work right now because I've introduced a bug with the label syntax, that treats semicolons incorrectly. I'm going to change the label syntax to a hash to avoid collisions:

```
# argdist.py -C 'p:c:nanosleep(struct timespec { time_t tv_sec; long tv_nsec; } *req):long:req->tv_nsec#nsec distribution'
```

In any case, replicating a complex structure is obviously counter-productive. I will add an -I switch that lets the user add include files. Something like this:

```
# argdist.py -C 'p::md_end_flush(struct bio *bio):int:bio->bi_private' -I 'linux/blkdev.h'
```

**Comment by goldshtn** [comment]
Added/fixed in https://github.com/goldshtn/bcc/commit/7983d6b61f83275471ee8fc4f2162facf4b36d91.

**Comment by goldshtn** [comment]
For reference, here's what the error message looks like now:

```
# ./argdist.py -I 'linux/blkdev.h' -C 'p:md_end_flush(struct bio *bio):int:bio->bi_private'
Traceback (most recent call last):
  File "./argdist.py", line 447, in <module>
    specifiers.append(Specifier("freq", specifier, args.pid))
  File "./argdist.py", line 178, in __init__
    self._validate_specifier()
  File "./argdist.py", line 174, in _validate_specifier
    "format") % parts[2])
  File "./argdist.py", line 158, in _bail
    (self.raw_spec, error))
ValueError: error parsing probe 'p:md_end_flush(struct bio *bio):int:bio->bi_private': function signature 'int' has an invalid format
```

**Comment by goldshtn** [comment]
I think this can be closed now? @brendangregg

**Comment by brendangregg** [comment]
Why does it print the traceback? That would be handy for a developer's verbose mode (-v), but by default I think the tools have better ease of use without dumping their own stack traces. eg:

```
$ ls -z
ls: invalid option -- 'z'
Try 'ls --help' for more information.
```

I'd be pretty surprised if ```ls``` gave me a traceback.

**Comment by goldshtn** [comment]
Good point. In my trace PR I've already implemented this behavior, and now fixing in argdist as well.

**Comment by goldshtn** [comment]
@brendangregg I believe this was fixed in #410, so I think this issue can be closed now. Thanks.

**Comment by brendangregg** [comment]
ok, thanks!
